src/Nethermind/Nethermind.Trie/TrieNode.Decoder.cs

```diff
@@ -36,26 +39,44 @@ public byte[] Encode(ITrieNodeResolver tree, TrieNode? item)
                     _ => throw new TrieException($"An attempt was made to encode a trie node of type {item.NodeType}")
                 };
             }
-
+            
+            [SkipLocalsInit]
             private static byte[] EncodeExtension(ITrieNodeResolver tree, TrieNode item)
             {
                 Debug.Assert(item.NodeType == NodeType.Extension,
                     $"Node passed to {nameof(EncodeExtension)} is {item.NodeType}");
                 Debug.Assert(item.Key is not null,
                     "Extension key is null when encoding");
 
-                byte[] keyBytes = item.Key.ToBytes();
                 TrieNode nodeRef = item.GetChild(tree, 0);
                 Debug.Assert(nodeRef is not null,
                     "Extension child is null when encoding.");
 
                 nodeRef.ResolveKey(tree, false);
 
+                HexPrefix hexPrefix = item.Key;
+                int hexLength = hexPrefix.ByteLength;
+
+                byte[]? rentedBuffer = hexLength > StackallocByteThreshold
+                    ? ArrayPool<byte>.Shared.Rent(hexLength)
+                    : null;
+
+                Span<byte> keyBytes = (rentedBuffer is null
+                    ? stackalloc byte[StackallocByteThreshold]
+                    : rentedBuffer).Slice(0, hexLength);
+
+                hexPrefix.CopyToSpan(keyBytes);
+
                 int contentLength = Rlp.LengthOf(keyBytes) + (nodeRef.Keccak is null ? nodeRef.FullRlp.Length : Rlp.LengthOfKeccakRlp);
                 int totalLength = Rlp.LengthOfSequence(contentLength);
                 RlpStream rlpStream = new(totalLength);
                 rlpStream.StartSequence(contentLength);
                 rlpStream.Encode(keyBytes);
+                if (rentedBuffer is not null)
+                {
+                    ArrayPool<byte>.Shared.Return(rentedBuffer);
+                }
+
```

Comment:
Would be be better in `finally` clause?

Reply:
`try` `catch` `finally` add somes overhead to methods as it causes the Jit to save more state to memory rather than working more with registers.

`ArrayPool<T>.Shared` isn't functionally impaired by not returning a rented array as it will just allocate a new one; so isn't really worth putting the extra overhead in to return the array in a `finally` as it effects the hot path.

If an exception is thrown that's very expensive, so the additional overhead of now having to allocate an extra array in the pool isn't that high by comparison.